File: acceptlanguageparser.go
package goacceptlanguageparser

import (
	"sort"
	"strconv"
	"strings"
)

type Language struct {
	name    string
	quality float64
}

type By func(l1, l2 *Language) bool

type languageSorter struct {
	languages []Language
	by        func(l1, l2 *Language) bool
}

func (by By) Sort(languages []Language) {
	ls := &languageSorter{
		languages: languages,
		by:        by,
	}
	sort.Sort(ls)
}

func (s *languageSorter) Len() int {
	return len(s.languages)
}

func (s *languageSorter) Swap(i, j int) {
	s.languages[i], s.languages[j] = s.languages[j], s.languages[i]
}

func (s *languageSorter) Less(i, j int) bool {
	return s.by(&s.languages[i], &s.languages[j])
}

func ParseAcceptLanguage(languages string, supported_languages []string) []string {
	langs := []Language{}
	browser_pref_langs := strings.Split(languages, ",")

	i := 0
	length := len(browser_pref_langs)

	for _, lang := range browser_pref_langs {
		// Format strings.
		lang = strings.Replace(strings.ToLower(strings.TrimSpace(lang)), "_", "-", 0)

		if lang != "" {

			// Search for quality.
			l := strings.SplitN(lang, ";", 2)
			var quality float64
			var err error

			if len(l) == 2 {
				q := l[1]

				if strings.HasPrefix(q, "q=") {
					q = strings.SplitN(q, "=", 2)[1]
					quality, err = strconv.ParseFloat(q, 64)

					if err != nil {
						// Default value (1) if quality is empty.
						quality = 1
					}
				}
			}

			// Use order of items if no quality is given.
			if quality == 0 {
				quality = float64(length - i)
			}

			language := l[0]

			// If supported languages are given, return only the langs that fit.
			if len(supported_languages) != 0 {
				for _, supported_lang := range supported_languages {
					if language == supported_lang {
						langs = append(langs, Language{language, quality})
						break
					}
				}
			} else {
				// If no supported language is given, return all langs.
				langs = append(langs, Language{language, quality})
			}

			i++
		}
	}

	// Sort in reverse order (quality descending).
	quality := func(l1, l2 *Language) bool {
		return l1.quality > l2.quality
	}

	By(quality).Sort(langs)

	// Filter quality string.
	langString := []string{}
	for _, lang := range langs {
		langString = append(langString, lang.name)
	}

	return langString

}

File: python/acceptlanguageparser.py
#!/usr/bin/python

# Copyright 2014 struktur AG. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

def parseAcceptLanguage(languages, supported_languages=None):

    langs = []
    browser_pref_langs = languages.split(',')

    i = 0
    length = len(browser_pref_langs)

    # parse quality strings and build a tuple
    # like ((float(quality), lang), (float(quality), lang))
    # which is sorted afterwards
    # if no quality string is given then the list order
    # is used as quality indicator
    for lang in browser_pref_langs:
        lang = lang.strip().lower().replace('_','-')
        if lang:
            l = lang.split(';', 2)
            quality = []

            if len(l) == 2:
                try:
                    q = l[1]
                    if q.startswith('q='):
                        q = q.split('=', 2)[1]
                        quality = float(q)
                except: pass

            if quality == []:
                quality = float(length-i)

            language = l[0]
            if supported_languages:
                if language in supported_languages:
                    # if allowed the add language
                    langs.append((quality, language))
            else:
                langs.append((quality, language))
            i = i + 1

    # sort and reverse it
    langs.sort()
    langs.reverse()

    # filter quality string
    langs = map(lambda x: x[1], langs)

    return langs

if __name__ == "__main__":

    actual = parseAcceptLanguage("en-US,en;q=0.8,de;q=0.6")
    print actual



